zebra-consensus/src/block/subsidy/general.rs

```diff
@@ -8,79 +8,101 @@ use zebra_chain::{
+            SLOW_START_SHIFT
+        ),
     }
 }
 
 /// `BlockSubsidy(height)` as described in [protocol specification §7.7][7.7]
 ///
 /// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
 pub fn block_subsidy(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
-    if height < SLOW_START_INTERVAL {
-        panic!("can't verify before block {}", SLOW_START_INTERVAL.0)
-    }
     let blossom_height = Blossom
         .activation_height(network)
         .expect("blossom activation height should be available");
+    let halving_div = halving_divisor(height, network);
 
-    let hd = halving_divisor(height, network);
-    if height >= blossom_height {
-        let scaled_max_block_subsidy = MAX_BLOCK_SUBSIDY / BLOSSOM_POW_TARGET_SPACING_RATIO;
-        // in future halvings, this calculation might not be exact
-        // in those cases, Amount division follows integer division, which truncates (rounds down) the result
-        Amount::try_from(scaled_max_block_subsidy / hd)
-    } else {
-        // this calculation is exact, because the halving divisor is 1 here
-        Amount::try_from(MAX_BLOCK_SUBSIDY / hd)
+    match height {
+        height if (height >= SLOW_START_INTERVAL && height < blossom_height) => {
+            // this calculation is exact, because the halving divisor is 1 here
+            Amount::try_from(MAX_BLOCK_SUBSIDY / halving_div)
+        }
+        height if (height >= blossom_height) => {
+            let scaled_max_block_subsidy = MAX_BLOCK_SUBSIDY / BLOSSOM_POW_TARGET_SPACING_RATIO;
+            // in future halvings, this calculation might not be exact
+            // in those cases, Amount division follows integer division, which truncates (rounds down) the result
+            Amount::try_from(scaled_max_block_subsidy / halving_div)
+        }
+        _ => unreachable!(
+            "unsupported block height: callers should handle blocks below {:?}",
+            SLOW_START_INTERVAL
+        ),
     }
 }
 
 /// `MinerSubsidy(height)` as described in [protocol specification §7.7][7.7]
 ///
 /// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+///
+/// `non_miner_reward` is the founders reward or funding stream value.
+/// If all the rewards for a block go to the miner, use `None`.
 #[allow(dead_code)]
-pub fn miner_subsidy(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
-    let canopy_height = Canopy
-        .activation_height(network)
-        .expect("canopy activation height should be available");
-    if height >= canopy_height {
-        panic!("Can't validate Canopy yet");
+pub fn miner_subsidy(
+    height: Height,
+    network: Network,
+    non_miner_reward: Option<Amount<NonNegative>>,
+) -> Result<Amount<NonNegative>, Error> {
+    if let Some(non_miner_reward) = non_miner_reward {
+        block_subsidy(height, network)? - non_miner_reward
+    } else {
+        block_subsidy(height, network)
```

Comment:
I like the founders reward dependency removal and how is handled with the new optional argument however in practice i think this will never happen.

Reply:
Here are the currently specified non-miner rewards:
* first halving: founders reward
* second halving: funding streams
* third halving onwards: None

See the funding streams ZIP for details:
https://zips.z.cash/zip-1014#abstract

We can't assume that funding streams will continue into the third halving - there would need to be a new ZIP for that change, and a community process.